hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorImpl.java

```diff
@@ -302,11 +348,53 @@ private void doFlush(boolean flushAll) throws InterruptedIOException,
         }
         asf = ap.submit(createTask(access));
       }
-      // DON'T do the wait in the try-with-resources. Otherwise, the undealt mutations won't
-      // be released.
-      asf.waitUntilDone();
-      if (asf.hasError()) {
-        errors.add(asf.getErrors());
+
+      if (flushAll || writeBufferSize == 0) {
+        // if we have setWriteBufferPeriodicFlushTimeoutMs we may have concurrent update
+        List<AsyncRequestFuture> waitList;
+        synchronized(asfList) {
```

Comment:
Oh, I see, BufferedMutator is thread safe, so multiple threads can enter this method at the same time and we want them to share the safe asfList, so it should be class member.